pkg/streamingpromql/operators/vector_scalar_binary_operation.go

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/compat"
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// VectorScalarBinaryOperation represents a binary operation between an instant vector and a scalar such as "<expr> + 2" or "3 * <expr>".
+type VectorScalarBinaryOperation struct {
+	Scalar                   types.ScalarOperator
+	Vector                   types.InstantVectorOperator
+	ScalarIsLeftSide         bool
+	Op                       parser.ItemType
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	start     int64 // Milliseconds since Unix epoch
+	end       int64 // Milliseconds since Unix epoch
+	interval  int64 // In milliseconds
+	stepCount int
+
+	opFunc vectorScalarBinaryOperationFunc
+
+	expressionPosition posrange.PositionRange
+	emitAnnotation     functions.EmitAnnotationFunc
+	scalarData         types.ScalarData
+	vectorIterator     types.InstantVectorSeriesDataIterator
+}
+
+type vectorScalarBinaryOperationFunc func(scalar float64, vectorF float64, vectorH *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error)
+
+func NewVectorScalarBinaryOperation(
+	scalar types.ScalarOperator,
+	vector types.InstantVectorOperator,
+	scalarIsLeftSide bool,
+	op parser.ItemType,
+	start int64,
+	end int64,
+	interval int64,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+) (*VectorScalarBinaryOperation, error) {
+	opFunc := arithmeticOperationFuncs[op]
```

Comment:
(nit) it's a bit confusing having opFunc here and b.opFunc being different types. Consider renamign this to `arithmeticOpFunc` or something.

Reply:
I've renamed it to `f` in 6779d85.